src/lib/GSENSOR/gsensor.cpp

```diff
@@ -0,0 +1,203 @@
+#ifdef HAS_GSENSOR
+#include "gsensor.h"
+#include "logging.h"
+
+#ifdef HAS_GSENSOR_STK8xxx
+#include "stk8baxx.h"
+STK8xxx stk8xxx;
+#endif
+
+int gensor_status = GSENSOR_STATUS_FAIL;
+
+#define DATA_SAMPLE_LENGTH  20
+#define QUIET_VARIANCE_THRESHOLD    0.0002f
+#define QUIET_AVERAGE_THRESHOLD     0.04f
+
+#define FLIPPED_ACTION_THRESHOLD  0.2f
+
+static float x_buffer[DATA_SAMPLE_LENGTH] = {0};
+static float y_buffer[DATA_SAMPLE_LENGTH] = {0};
+static float z_buffer[DATA_SAMPLE_LENGTH] = {0};
+static int sample_counter = 0;
+float x_average = 0;
+float y_average = 0;
+float z_average = 0;
+
+static int motion_event_counter = 0;
+
+#ifdef HAS_THERMAL
+extern bool is_smart_fan_control;
+uint32_t smart_fan_start_time = 0;
+#define SMART_FAN_TIME_OUT 5000
+#endif
+
+ICACHE_RAM_ATTR void handleGsensorInterrupt() {
+    motion_event_counter++;
+}
+
+void Gsensor::init()
+{
+    Wire.begin(GPIO_PIN_SDA, GPIO_PIN_SCL);
+
+    uint8_t id = -1;
+#ifdef HAS_GSENSOR_STK8xxx
+    id = stk8xxx.STK8xxx_Initialization();
+#endif    
+    if(id == -1)
+    {
+        ERRLN("Gsensor failed!");
+    }
+    else
+    {
+        INFOLN("Gsensor OK with chipid = %x!", id);
+        gensor_status = GSENSOR_STATUS_NORMAL;
+    }
+
+    if(gensor_status == GSENSOR_STATUS_NORMAL)
+    {
+#ifdef HAS_GSENSOR_STK8xxx        
+       stk8xxx.STK8xxx_Anymotion_init();
+       pinMode(GPIO_PIN_GSENSOR_INT,INPUT_PULLUP);
+       attachInterrupt(digitalPinToInterrupt(GPIO_PIN_GSENSOR_INT), handleGsensorInterrupt, FALLING);
+#endif       
+    }
+
+    system_state = GSENSOR_SYSTEM_STATE_MOVING;
+    is_flipped = false;
+}
+
+float get_data_average(float *data, int length)
+{
+    float average = 0;
+    for(int i=0;i<length;i++)
+    {
+        average += *(data+i);
+    }
+    average = average / length;
+    return average;  
+}
+
+float get_data_variance(float average, float *data, int length)
```

Comment:
Are functions like `get_data_average` and `get_data_variance` better being in helper.h?

Reply:
What is mean, where is the helper.h?

Reply:
`lib/helpers/helper.h`

Reply:
see the helper.h,  it is not suitable for this function

Reply:
@JyeSmith did you mean move the functions to the lib/helpers/helpers.h?